<code>Enumerable.Sum()</code> always executes addition in a <code>checked</code> context, so an <code>OverflowException</code> will be thrown if the value exceeds <code>MaxValue</code> even if an <code>unchecked</code> context was specified. Using an <code>unchecked</code> context anyway represents a misunderstanding of how <code>Sum</code> works.

This rule raises an issue when an <code>unchecked</code> context is specified for a <code>Sum</code> on integer types.


== Noncompliant Code Example

----
void Add(List<int> list)
{
  int d = unchecked(list.Sum());  // Noncompliant

  unchecked 
  {
    int e = list.Sum();  // Noncompliant
  }
}
----


== Compliant Solution

----
void Add(List<int> list)
{
  int d = list.Sum();

  try 
  {
    int e = list.Sum();
  } 
  catch (System.OverflowException e) 
  {
    // exception handling...
  }
}
----


== Exceptions

When the <code>Sum()</code> call is inside a <code>try-catch</code> block, no issues are reported.

----
void Add(List<int> list)
{
  unchecked 
  {
    try 
    {
      int e = list.Sum();
    } 
    catch (System.OverflowException e) 
    {
      // exception handling...
    }
  }
}
----


